enumerables.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ModuleLength,
+module Enumerable
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_each
+    i = 0
+    return to_enum unless block_given?
+
+    while i < size
+      yield(to_a[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    i = 0
+    return to_enum unless block_given?
+
+    while i < size
+      yield(to_a[i], i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    if block_given?
+      arr = []
+      my_each do |i|
+        arr << i if yield(i)
+      end
+    end
+    arr
+  end
+
+  def my_all?(input = nil)
+    my_each do |i|
+      return false if block_given? && !yield(i)
+
+      if !block_given? && input.nil?
+        return false unless i
+      elsif input
+        return false unless check_input(i, input)
+      end
+    end
+    true
+  end
+
+  def my_any?(input = nil)
+    my_each do |i|
+      return true if block_given? && yield(i)
+
+      if !block_given? && input.nil?
+        return true if i
+      elsif !block_given? && input
+        return true if check_input(i, input)
+      end
+    end
+    false
+  end
+
+  def my_none?(input = nil)
+    my_each do |i|
+      return false if block_given? && yield(i)
+
+      if !block_given? && input.nil?
+        return false if i
+      elsif !block_given? && input
+        return false if check_input(i, input)
+      end
+    end
+    true
+  end
+
+  def my_count(input = nil)
+    counter = 0
+    my_each do |i|
+      if block_given?
+        counter += 1 if yield(i)
+      elsif i && input.nil?
+        counter += 1 unless block_given?
+      elsif check_input(i, input)
+        counter += 1 unless block_given?
+      end
+    end
+    counter
+  end
+
+  def my_map
+    arr = []
+    my_each do |i|
+      return to_enum unless block_given?
+
+      arr << yield(i) || arr << proc.call(i) if block_given?
+    end
+    arr
+  end
+
+  def my_inject(start = nil, symbol = nil)
+    if block_given?
+      memo = start if start
+      if start.nil?
```

Comment:
- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method


```
irb(main):249:0> ok = [1,2,3,4,5]
=> [1, 2, 3, 4, 5]
irb(main):250:0> ok.my_inject(:+)
=> 15
irb(main):251:0> ok.my_inject(:+)
=> 14
irb(main):252:0> ok.my_inject(:+)
=> 12
irb(main):253:0> ok.my_inject(:+)
=> 9
irb(main):254:0> ok.inject(:+)
=> 5
irb(main):255:0> ok.inject(:+)
=> 5
irb(main):256:0> ok.inject(:+)
=> 5
irb(main):257:0> ok.inject(:+)
=> 5
irb(main):258:0> ok.inject(:+)
=> 5
irb(main):259:0> ok.inject(:+)
=> 5
irb(main):260:0> ok.my_inject(:+)
=> 5
irb(main):261:0> ok.my_inject(:+)
=> nil
irb(main):262:0> ok.my_inject(:+)
=> nil
``` 

- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block


```
irb(main):270:0> arr = [1,2,3,4,5]
=> [1, 2, 3, 4, 5]
irb(main):271:0> arr.inject {|n,j| n+j}
=> 15
irb(main):272:0> arr.inject {|n,j| n+j}
=> 15
irb(main):273:0> arr.inject {|n,j| n+j}
=> 15
irb(main):274:0> arr.inject {|n,j| n+j}
=> 15
irb(main):275:0> arr.inject {|n,j| n+j}
=> 15
irb(main):276:0> arr.inject {|n,j| n+j}
=> 15
irb(main):277:0> arr.my_inject {|n,j| n+j}
=> 15
irb(main):278:0> arr.my_inject {|n,j| n+j}
=> 14
irb(main):279:0> arr.my_inject {|n,j| n+j}
=> 12
irb(main):280:0> arr.my_inject {|n,j| n+j}
=> 9
irb(main):281:0> arr.my_inject {|n,j| n+j}
=> 5
irb(main):282:0> arr.inject {|n,j| n+j}
=> nil
irb(main):283:0> arr.inject {|n,j| n+j}
=> nil

```

Reply:
Thanks for the review! I was just testing the two cases you told me and I'm getting the right answer. 

`irb(main):002:0> [1, 2, 3, 4, 5].my_inject {|n,j| n+j}
=> 15
irb(main):003:0> [1, 2, 3, 4, 5].inject {|n,j| n+j}
=> 15
irb(main):004:0> [1, 2, 3, 4, 5].my_inject(:+)
=> 15
irb(main):005:0> [1, 2, 3, 4, 5].inject(:+)
=> 15`

Reply:
Hi!, try saving the array into a variable and repeating the procedure a few times

Reply:
```
arr = [1,2,3,4,5]
```

and then 

```
irb(main):274:0> arr.inject {|n,j| n+j}
=> 15
irb(main):275:0> arr.inject {|n,j| n+j}
=> 15
irb(main):276:0> arr.inject {|n,j| n+j}
=> 15
irb(main):277:0> arr.my_inject {|n,j| n+j}
=> 15
irb(main):278:0> arr.my_inject {|n,j| n+j}
=> 14
irb(main):279:0> arr.my_inject {|n,j| n+j}
=> 12
irb(main):280:0> arr.my_inject {|n,j| n+j}
=> 9
irb(main):281:0> arr.my_inject {|n,j| n+j}
=> 5
```